core/src/Kokkos_Cuda.hpp

```diff
@@ -249,12 +250,12 @@ class Cuda {
   static const char* name();
 
   inline Impl::CudaInternal* impl_internal_space_instance() const {
-    return m_space_instance;
+    return m_space_instance.get();
   }
   uint32_t impl_instance_id() const noexcept { return 0; }
 
  private:
-  Impl::CudaInternal* m_space_instance;
+  std::shared_ptr<Impl::CudaInternal> m_space_instance;
```

Comment:
This adds overhead _everywhere_ in every application, even if they don't use streams.  Also, anyone I'm not sure we're comfortable with that. Also, any downstream users who are (probably by accident, but still) copying execution space instances on the device will now get `SEGFAULT`s in code that previously worked fine.

Reply:
alternatively I would just reference count it explicitly (i.e. add an int, atomic increment it in copy and decrement it destructor (though that makes it non-trivial copyable)

Reply:
no, if they try to copy. `std::shared_ptr` has a nontrivial copy constructor that CUDA would try to invoke on the device. Since a lot of our users ignore warnings, it would cause a crash.

Reply:
> (though that makes it non-trivial copyable)

It's non-trivially copyable anyway. `std::shared_ptr` isn't trivially copyable.

Reply:
maybe we should add a `desul::scoped_shared_ptr<AtomicScope>`